gh-117657: Fixes a few small TSAN issues in dictobject

DinoV opened the pull request:
There's a few spots in the dict implementation where we're not using atomics to access or update the dictionary size.  This updates those so that we're using relaxed atomics on them.

The `dictobject.h` case is a little special because we don't have the `FT` macros available, so it's special cased with an ifdef.

<!-- gh-issue-number: gh-117657 -->
* Issue: gh-117657
<!-- /gh-issue-number -->

DinoV committed:
Fixup TSAN errors for dict

mpage commented on Tools/tsan/suppressions_free_threading.txt:
```diff
@@ -27,9 +27,6 @@ race:_PyObject_GC_SET_SHARED
 race:_PyObject_GC_TRACK
 race:_PyType_HasFeature
 race:assign_version_tag
-race:compare_unicode_unicode
```
This seems like it should be unrelated to accessing `ma_used`. IIRC it looked like a race on the unicode object's hash value.

DinoV replied:
There's not a fix for it, but I'm not seeing any races reported against it after disabling it (it's a helper function in `dictobject.c`)

DinoV committed:
Fix more dict races:
    Keys need atomic relaxed loads as they can be assigned from another thread
    Entries need atomic release assigns so the values they point to are seen
    Reading managed dict needs acquire so that the memory of a newly created and published dictionary is visible
